llvm/lib/Transforms/IPO/DynUntrustedAllocPost.cpp

```diff
@@ -0,0 +1,250 @@
+
+using namespace llvm;
+
+namespace {
+class IDGenerator {
+  unsigned int id;
+
+public:
+  IDGenerator() : id(0) {}
+
+  ConstantInt *getConstID(Module &M) {
+    return llvm::ConstantInt::get(IntegerType::getInt64Ty(M.getContext()),
+                                  id++);
+  }
+
+  ConstantInt *getDummyID(Module &M) {
+    return llvm::ConstantInt::get(IntegerType::getInt64Ty(M.getContext()), -1);
+  }
+};
+
+static IDGenerator IDG;
+
+struct FaultingSite {
+  uint64_t uniqueID;
+  int64_t pkey;
+};
+
+class DynUntrustedAllocPost : public ModulePass {
+public:
+  static char ID;
+
+  DynUntrustedAllocPost() : ModulePass(ID) {
+    initializeDynUntrustedAllocPostPass(*PassRegistry::getPassRegistry());
+  }
+  virtual ~DynUntrustedAllocPost() = default;
+
+  bool runOnModule(Module &M) override {
+    // Post inliner pass, iterate over all functions and find hook CallSites.
+    // Assign a unique ID in a deterministic pattern to ensure UniqueID is
+    // consistent between runs.
+    assignUniqueIDs(M);
+    fixFaultedAllocations(M, getFaultingAllocList());
+
+    return true;
+  }
+
+  bool fromJSON(const llvm::json::Value Alloc, FaultingSite &F) {
+    llvm::json::ObjectMapper O(Alloc);
+    int64_t temp;
+    bool temp_result = O.map("id", temp);
+    if (temp < 0) {
+      return false;
+    }
+    F.uniqueID = static_cast<uint64_t>(temp);
+    return O && temp_result && O.map("pkey", F.pkey);
+  }
+
+  std::set<FaultingSite> getFaultingAllocList() {
+    // TODO : Somehow we need to get the path to the .json file containing faults.
+    std::set<FaultingSite> fault_set;
+
+    return fault_set;
+  }
+
+  int getArgIndexForPatch(Function *hook) {
+    assert(hook && "Nullptr: Invalid function to hook");
+    auto hookFns = getHookFuncs(*hook->getParent());
+    if (hook == hookFns[0]) {
+      return 2;
+      // allocHook
+    } else if (hook == hookFns[1]) {
+      return 4;
+      // mallocHook
+    } else if (hook == hookFns[2]) {
+      // deallocHook
+      return 2;
+    }
+    return -1;
+  }
+
+  SmallVector<Function *, 3> getHookFuncs(Module &M) {
+    std::string hookFuncNames[3] = {"allocHook", "reallocHook", "deallocHook"};
+    SmallVector<Function *, 3> hookFns;
+    for (auto hookName : hookFuncNames) {
+      Function *F = M.getFunction(hookName);
+      if (!F)
+        continue;
+
+      hookFns.push_back(F);
+    }
+    return hookFns;
+  }
+
+  static bool funcSort(Function *F1, Function *F2) {
+    return F1->getName().data() > F2->getName().data();
+  }
+
+  void assignUniqueIDs(Module &M) {
+    std::vector<Function *> WorkList;
+    for (Function &FRef : M) {
+      Function *F = &FRef;
+      if (!F)
+        continue;
+
+      if (F->isDeclaration())
+        continue;
+
+      WorkList.push_back(F);
+    }
+
+    std::sort(WorkList.begin(), WorkList.end(), funcSort);
+
+    LLVM_DEBUG(errs() << "Search for modified functions!\n");
+
+    for (auto *F : llvm::reverse(WorkList)) {
```

Comment:
why reverse the worklist? if you need reverse sorted, you can do that by either reversing the comparator in the sort or by using a reverse iterator.

But I can't see a reason why the list should be walked in a reverse order at all? the `AllocID`'s just need to be stable, nothing more. I don't see a reason why starting at the end of the alphabet is beneficial.

Reply:
Removed reversing list.